src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -363,27 +365,38 @@ internal override ValueTask<int> ReadAsync(Memory<byte> destination, Cancellatio
                 NetEventSource.Info(_state, $"{TraceId()} Stream reading into Memory of '{destination.Length}' bytes.");
             }
 
-            ReadState readState;
-            long abortError = -1;
-            bool canceledSynchronously = false;
+            ReadState initialReadState;  // value before transitions
+            long abortError;
+            bool preCanceled = false;
 
             lock (_state)
             {
-                readState = _state.ReadState;
+                initialReadState = _state.ReadState;
                 abortError = _state.ReadErrorCode;
 
-                if (readState != ReadState.PendingRead && cancellationToken.IsCancellationRequested)
+                // Failure scenario: pre-canceled token. Transition: any -> Aborted
+                // PendingRead state indicates there is another concurrent read operation in flight
+                // which is forbidden, so it is handled separately
+                if (initialReadState != ReadState.PendingRead && cancellationToken.IsCancellationRequested)
                 {
-                    readState = ReadState.Aborted;
+                    initialReadState = ReadState.Aborted;
                     _state.ReadState = ReadState.Aborted;
-                    canceledSynchronously = true;
+                    preCanceled = true;
```

Comment:
Why don't directly return `new OperationCanceledException(cancellationToken)` here and get rid of `preCanceled` local var. We wouldn't even need to change the `initialReadState` if I'm reading the code correctly.

Reply:
To do all the work on exceptions in one place and outside of the lock